cpu/stm32f1/periph/rtc.c

```diff
@@ -0,0 +1,265 @@
+
+    /* disable backup domain write protection */
+    PWR->CR &= ~PWR_CR_DBP;
+}
+
+static time_t _rtc_get_time(void)
+{
+    return (RTC->CNTH << 16) | RTC->CNTL;
+}
+
+static void _rtc_set_time(time_t counter_val)
+{
+    _rtc_enter_config_mode();
+    RTC->CNTH = (counter_val & 0xffff0000) >> 16;
+    RTC->CNTL = counter_val & 0x0000ffff;
+    _rtc_exit_config_mode();
+}
+
+void rtc_init(void)
+{
+    /* save current time if RTC already works */
+    bool is_rtc_enable = _is_rtc_enable();
+    time_t cur_time = 0;
+    if (is_rtc_enable) {
+        cur_time = _rtc_get_time();
+    }
+
+    /* config RTC */
+    _rtc_config();
+
+    /* restore current time after config */
+    if (is_rtc_enable) {
+        _rtc_set_time(cur_time);
+    }
+}
+
+int rtc_set_time(struct tm *time)
+{
+    time_t timestamp = mktime(time);
+
+    if (timestamp == -1) {
+        return -1;
+    }
+
+    _rtc_set_time(timestamp);
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+int rtc_get_time(struct tm *time)
+{
+    time_t timestamp = _rtc_get_time();
+    localtime_r(&timestamp, time);
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+static void _rtc_enable_alarm(void)
+{
+    /* clear alarm flag */
+    RTC->CRL &= ~RTC_CRL_ALRF;
+
+    _rtc_enter_config_mode();
+    RTC->CRH |= (RTC_CRH_ALRIE);
+    _rtc_exit_config_mode();
+}
+
+static void _rtc_disable_alarm(void)
+{
+    _rtc_enter_config_mode();
+    RTC->CRH &= ~(RTC_CRH_ALRIE | RTC_CRH_SECIE);
+    _rtc_exit_config_mode();
+}
+
+static time_t _rtc_get_alarm_time(void)
+{
+    return (RTC->ALRH << 16) | RTC->ALRL;
+}
+
+static void _rtc_set_alarm_time(time_t alarm_time)
+{
+    _rtc_enter_config_mode();
+    RTC->ALRL = (alarm_time & 0x0000ffff);
+    RTC->ALRH = (alarm_time & 0xffff0000) >> 16;
+    _rtc_exit_config_mode();
+}
+
+int rtc_set_alarm(struct tm *time, rtc_alarm_cb_t cb, void *arg)
+{
+    /* enable global RTC interrupt */
+    if (cb && !NVIC_GetEnableIRQ(RTC_IRQn)) {
+        NVIC_EnableIRQ(RTC_IRQn);
+        DEBUG("%s enable RTC IRQ\n", __func__);
+    }
+
+    /* save callback and argument */
+    isr_ctx.cb = cb;
+    isr_ctx.arg = arg;
+
+    /* set wakeup time */
+    time_t timestamp = mktime(time);
+
+    if (timestamp == -1) {
+        return -2;
+    }
+
+    _rtc_disable_alarm();
+    _rtc_set_alarm_time(timestamp);
+
+    /* enable Alarm */
+    _rtc_enable_alarm();
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+int rtc_get_alarm(struct tm *time)
+{
+    time_t timestamp = _rtc_get_alarm_time();
+    localtime_r(&timestamp, time);
+
+    DEBUG("%s timestamp=%"PRIu32"\n", __func__, (uint32_t)timestamp);
+
+    return 0;
+}
+
+static void _rtc_clear_isr_ctx(void)
+{
+    isr_ctx.cb = NULL;
+    isr_ctx.arg = NULL;
+}
+
+void rtc_clear_alarm(void)
+{
+    _rtc_disable_alarm();
+    _rtc_set_alarm_time(0);
+
+    _rtc_clear_isr_ctx();
```

Comment:
I think you don't need this here either.

Reply:
`_rtc_clear_isr_ctx`?
On other processor efm32/rtc, esp32/rtc and other callback is cleared.
I think clear need.

Reply:
just "because others are doing it" is never good reasoning 😉 

The alarm can only be enabled again by `rtc_set_alarm()` where those are overwritten anyway.
It does no harm, but it's a waste of instructions and adds needless clutter to the code.

Reply:
You can drop the `_rtc_set_alarm_time(0)` too.

Reply:
See other `rtc_clear_alarm(void)`. All contain callback clear.

Reply:
> You can drop the `_rtc_set_alarm_time(0)` too.

Function name - `rtc_clear_alarm`, not `rtc_disable_alarm`.
After `rtc_clear_alarm` we can run `rtc_get_alarm` and see that alarm not clear - only disable

Reply:
`lpc2387` and  `cc430` do not… 
But yea there seems to be some strange cargo cult going on.

Well keep it if you want.
But the `_rtc_set_alarm_time(0)` won't help you if the counter wraps around and `_rtc_disable_alarm()` was not working for some reason.

Reply:
> `lpc2387` and `cc430` do not…
> But yea there seems to be some strange cargo cult going on.
> 
> Well keep it if you want.
> But the `_rtc_set_alarm_time(0)` won't help you if the counter wraps around and `_rtc_disable_alarm()` was not working for some reason. (Which would be the only reason why you would want to delete the callback)

I found case.
We set time to 15:20:59, and set alarm 15:21:10. Alarm called.
At cur time 15:21:20 - set time 15:20:59.
Now we have: time 15:20:59. Alarm 15:21:10 (`rtc_get_alarm` return). But alarm not called (is disable)
I think that after callback we need clear alarm time!

Reply:
> just "because others are doing it" is never good reasoning wink
> 
> The alarm can only be enabled again by `rtc_set_alarm()` where those are overwritten anyway.
> It does no harm, but it's a waste of instructions and adds needless clutter to the code.

Ok. I can argee with `clear_isr_ctx`. `isr_ctx` not visible outside. But alarm time visible on `rtc_get_alarm`.

Reply:
Hm I would have expected `rtc_get_alarm()` to return -1 if no alarm is set, but this doesn't seem to be the case.

So returning an alarm in the past might be the next best thing.
IMHO the RTC API needs some overhaul…

Reply:
> Hm I would have expected `rtc_get_alarm()` to return -1 if no alarm is set, but this doesn't seem to be the case.
> 
> So returning an alarm in the past might be the next best thing.
> IMHO the RTC API needs some overhaul…

If clear alarm on callback, then test periph_rtc work wrong:
`    puts("Alarm!");`

`    if (++cnt < REPEAT) {`
`        struct tm time;`
`        rtc_get_alarm(&time);`
`        inc_secs(&time, PERIOD);`
`        rtc_set_alarm(&time, cb, NULL);`
`    }`
`rtc_get_alarm` - return zero time and next 3 alarm not run.
If `rtc_get_alarm` return `-1` when alarm is disable (on callback too) - same problem. Alarm run only 1 times. Returned value `rtc_get_alarm` not check in test. I think replace `rtc_get_alarm` to `rtc_get_time` - good things.

Reply:
Yea like I said, I think the API needs a bit of rework to be less ambiguous.
Anyway, your implementation works and should not be help up over different interpretations of the unclear RTC API.